internal/workflow/preprocessing.go

```diff
@@ -47,50 +55,109 @@ func (w *PreprocessingWorkflow) Execute(
 		e = temporal.NewNonRetryableError(fmt.Errorf("error calling workflow with unexpected inputs"))
 		return nil, e
 	}
+	result.RelativePath = params.RelativePath
 
 	localPath := filepath.Join(w.sharedPath, filepath.Clean(params.RelativePath))
 
 	// Identify SIP.
+	identifySIPEvent := newEvent(ctx, "Identify SIP structure")
 	var identifySIP activities.IdentifySIPResult
 	e = temporalsdk_workflow.ExecuteActivity(
 		withLocalActOpts(ctx),
 		activities.IdentifySIPName,
 		&activities.IdentifySIPParams{Path: localPath},
 	).Get(ctx, &identifySIP)
+
 	if e != nil {
-		return nil, e
+		identifySIPEvent.fail(
+			ctx,
+			"Content error: SIP structure could not be identified. Please ensure that SIP conforms to one of the following structures: %s",
+			strings.Join(enums.SIPTypeNames(), ", "),
+		)
+	} else {
+		identifySIPEvent.succeed(ctx, "SIP structure identified: %s", identifySIP.SIP.Type.String())
+	}
+	result.EventLog.Append(identifySIPEvent.Event)
```

Comment:
Right now the `IdentifySIP` activity can really only return a system error or an identified `sip.SIP`.  If I return a system error directly from the preprocessing workflow (i.e. `return nil, err`) then no event data will be returned to the parent workflow because Temporal workflows can not return both a result and an error (ref: https://community.temporal.io/t/failing-a-workflow-and-returning-data/6620/5).

So I think what I actually need to do here is:

1. If IdentifySIP returns an error, then record the error information in the identifySIPEvent and return that data in the `PreprocessingWorkflowResult` (e.g. `return PreprocessingWorkflowResult{eventLog: ...}, nil`)
2. If IdentifySIP doesn't return an error, update the identifySIPEvent appropriately, and continue the preprocessing workflow

Reply:
I read the conversation you had with Sevein in Slack. I was just thinking on changing that activity as you did with the others to follow the same pattern in workflow, at some point it may return a system error instead of an identification error. Also, not a big deal.